File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:

int helper(vector<int> &nums, int target, int index)
{
   if (index == 0)
    {
        if (target == nums[0] && target == -nums[0])
            return 2;
        if (target == nums[0] || target == -nums[0])
            return 1;
        return 0;
    }

    // plus
    int left = helper(nums, target - nums[index], index - 1);
    // minus
    int right = helper(nums, target + (nums[index]), index - 1);
    return left + right;
}

int findTargetSumWays(vector<int> &nums, int target)
{
    int n = nums.size();
    
    return helper(nums, target, n - 1);
}
};

File: 0494-target-sum/0494-target-sum.java
class Solution {
 Map<String, Integer> dp;

    int helper(int[] nums, int index, int target) {
        int n = nums.length;

        if (index == n) {
            if (target == 0)
                return 1;
            return 0;
        }
        String key = index + "|" + target;
        if (dp.containsKey(key))
            return dp.get(key);
        // pick,,,
        int pick = helper(nums, index + 1, target - nums[index]);
        // not pick...
        int notPick = helper(nums, index + 1, target + nums[index]);
        dp.put(key, pick + notPick);
        return pick + notPick;
    }

    public int findTargetSumWays(int[] nums, int target) {

        int n = nums.length;
        dp = new HashMap<>();
        return helper(nums, 0, target);
    }
}